### PR TITLE
SNOW-1571136 Add `network tcp-check` command

### PR DESCRIPTION
**Pull request by sfc-gh-ikryvanos**
This PR introduces a new service `network` with new subcommand `tcp-check` to check TCP connectivity from sansshell target(s) to the given host and port

# TODO:
- [x] cover with unit and integration tests

# Is it was tested
Yes, on local
```
$ sanssh --targets=localhost network tcp-check 500.500.500:8080
- From target localhost:50042 to 500.500.500:8080, status: Failed - Connection refused
Connectivity check completed.

$ sanssh --targets=localhost network tcp-check google.com:80 --timeout 5
- From target localhost:50042 to google.com:80, status: Succeed
Connectivity check completed.

$ sanssh --targets=localhost network tcp-check 500.500:1 --timeout 5
- From target localhost:50042 to 500.500:1, status: Failed - No such host
Connectivity check completed.

$ sanssh --targets=localhost network tcp-check 10.0.5.3:1 --timeout 5
- From target localhost:50042 to 10.0.5.3:1, status: Failed - Connection timed out
Connectivity check completed.
```

## How its look like
image

### CODE REVIEW

File: services/network/client/infrastructure/input/tcp-check.controller.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		cliUtils.ColoredFprintf(os.Stderr, cliUtils.RED_TEXT_COLOR, "--host flag is required.\n")
+		return subcommands.ExitUsageError
+	}
+
+	if err := validator.IsValidPort(p.port); err != nil {
+		cliUtils.ColoredFprintf(os.Stderr, cliUtils.RED_TEXT_COLOR, "Invalid port number: %s\n", err.Error())
+		return subcommands.ExitUsageError
+	}
+
+	preloader := cliUtils.NewDotPreloader("Waiting for results from remote machines")
```
Comment (sfc-gh-srhodes) [inline services/network/client/infrastructure/input/tcp-check.controller.go:68]: This part feels like it should be in sanssh-level code that applies to all subcommands and not in an individual action. Maybe you can do a separate PR that does it?

File: services/network/client/infrastructure/input/tcp-check.controller.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+}
+func (*TCPCheckCmd) Usage() string {
+	return `tcp-check --host <host> --port <port>:
+    Makes tcp connectivity check from the remote machine to specified server.
+`
+}
+
+func (p *TCPCheckCmd) SetFlags(f *flag.FlagSet) {
+	f.StringVar(&p.host, "host", "", "Host to check connectivity from remote machine")
+	f.IntVar(&p.port, "port", 8080, "Port to check connectivity from remote machine")
```
Comment (sfc-gh-srhodes) [inline services/network/client/infrastructure/input/tcp-check.controller.go:51]: Uintvar?

Also, maybe make it a default of 0 and fail if set to 0? A default of 8080 doesn't seem useful.

If an arg is always required, we often get it from f.Args() instead of a flag value.

File: services/network/server/server.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	timeout := req.GetTimeout()
+
+	if err := validator.IsValidPortUint32(rawPort); err != nil {
+		logger.Error(err, "Invalid port value")
+		recorder.CounterOrLog(ctx, s.tcpCheckFailureCounter, 1)
+		return nil, status.Errorf(codes.Internal, "Invalid port value: %s")
+	}
+	port := uint8(rawPort)
+
+	usecase := app.NewTCPCheckUsecase(&infraOutput.TCPClient{})
```
Comment (sfc-gh-srhodes) [inline services/network/server/server.go:56]: I might be missing something in why you chose this design.

What's the advantage of the TCPCheckUsecase abstraction? It's used exactly once and it doesn't do much more than wrap a single net.Dialer call.

**Review by sfc-gh-srhodes — CHANGES_REQUESTED** [review]

File: services/network/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+```bash
+sanssh --targets <remote-machine-hosts> network tcp-check --host <host> --port <port>
+```
+Where:
+- `<remote-machine-hosts>` is a list of remote machine hosts
+- `<host>` is the host to check
+- `<port>` is the port to check
+
+## Module architecture
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/README.md:18]: Actually it not the same. Other modules mostly defines client and server in single file. If you and @sfc-gh-elinardi agreed I would prefer to use this architecture for all modules

File: services/network/server/server.go (hunk 3 above)
Comment (sfc-gh-ikryvanos) [reply services/network/server/server.go:56]: It was made to separate api adapter (grpc controller) and application logic to improve testability. In this particular case it is not so visible because action is super simple

File: services/network/client/infrastructure/input/tcp-check.controller.go (hunk 2 above)
Comment (sfc-gh-ikryvanos) [reply services/network/client/infrastructure/input/tcp-check.controller.go:51]: > Uintvar?

Agreed

> Also, maybe make it a default of 0 and fail if set to 0? A default of 8080 doesn't seem useful.

Agreed

> If an arg is always required, we often get it from f.Args() instead of a flag value.

I uses flags because IMAO it gives more transparency when you read/share/use commands. It just came to my mind I could use `<host>:<port>` string as position argument and it would be much easier for end users to use

File: services/network/client/infrastructure/input/tcp-check.controller.go (hunk 1 above)
Comment (sfc-gh-ikryvanos) [reply services/network/client/infrastructure/input/tcp-check.controller.go:68]: Not sure. Controller prints result from each host as it get results and of keeps this preloaded very last line until all targets not gives its reply. IMAO, how to handle progress-bar/pre-loader it hardly depends on what commands do on local machine and how it handles server outputs, extract this logic on a sanssh-level could be either tricky or may not work in all cases

File: services/network/client/application/tcp-check.usecase.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+package application
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	pb "github.com/Snowflake-Labs/sansshell/services/network"
+)
+
+type TCPCheckResult struct {
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/client/application/tcp-check.usecase.go:25]: it makes app hardly depends on chosen grpc generation library. Any significant changes in library or proto files will require a lot of changes. If we use own structures we could cover usecase with unit tests and be sure that cli controller would work if it works

File: services/network/server/infrastructure/output/tcp.client.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+	timeout := time.Duration(timeoutSeconds) * time.Second
+
+	dialer := net.Dialer{Timeout: timeout}
+	conn, err := dialer.DialContext(ctx, "tcp", hostToCheck)
+
+	if err != nil {
+		var failReason string
+		if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+			failReason = "Connection timed out"
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/server/infrastructure/output/tcp.client.go:28]: It would cause proto changes each time we want add new message from backend. In current implementation we could add as much "failReason" on server as we want without releasing new proto schema 

Also not sure how GRPC would handle when we extend proto and deploy new version to the server, but client would use old version, potentially it could cause parsing error. I could check if you still thing it is good idea

File: services/network/network.proto [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  // The port to check
+  uint32 port = 2;
+  // The timeout in seconds
+  uint32 timeout = 3;
+}
+
+// TCPCheckReply describes the result of a rpc
+message TCPCheckReply {
+  bool ok = 1;
+  optional string failReason = 2;
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/network.proto:42]: Good catch, thank you

File: services/util/cli/utill.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+	"fmt"
+	"io"
+)
+
+var restoreFormatingCode = "\033[0m"
+var RED_TEXT_COLOR = "\033[31m"
+var GREEN_TEXT_COLOT = "\033[32m"
+var YELLOW_TEXT_COLOR = "\033[33m"
+
+func ColoredFprintf(w io.Writer, color string, format string, a ...any) (n int, err error) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/util/cli/utill.go:13]: image

In VIM its like this. Do you think its critical?

File: services/network/client/application/tcp-check.usecase.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sfc-gh-srhodes) [reply services/network/client/application/tcp-check.usecase.go:25]: > it makes app hardly depends on chosen grpc generation library. Any significant changes in library or proto files will require a lot of changes. If we use own structures we could cover usecase with unit tests and be sure that cli controller would work if it works

Why would depending on the grpc generation be a problem? The grpc libraries have interface stability guarantees and the protos are part of SansShell's public API and need to keep backwards compatibility as part of that API.

File: services/network/server/infrastructure/output/tcp.client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sfc-gh-srhodes) [reply services/network/server/infrastructure/output/tcp.client.go:28]: As long as you do https://protobuf.dev/programming-guides/dos-donts/#unspecified-enum, backwards-compatibility for enums in public interfaces works fine.

File: services/network/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfc-gh-srhodes) [reply services/network/README.md:18]: Fair, the "application" and "infrastructure" divisions are new.

I'd still rather keep the documentation within a module limited to things specific to that module. If you want to use this architecture for more modules, I'd rather put the docs in a common place and link to them from here.

File: services/network/client/infrastructure/input/tcp-check.controller.go (hunk 1 above)
Comment (sfc-gh-srhodes) [reply services/network/client/infrastructure/input/tcp-check.controller.go:68]: I think uniformity across commands is useful. I don't mind changing how our outputting works, but I'd rather do it for all commands and not a subset of them.

File: services/util/cli/utill.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sfc-gh-srhodes) [reply services/util/cli/utill.go:29]: I'd prefer to not have terminal codes show up in file output.

I've seen other similar implementations deal with this by using different behavior based on https://pkg.go.dev/golang.org/x/term#IsTerminal and similar.

File: services/network/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/README.md]: done

File: services/network/README.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
+# Network
+This directory contains the network services. There is should be defined network related commands like `ping`, `traceroute`, `netstat`, etc.
+What is not part of network service? Commands related to protocols build on top of network layer like `http`, `ftp`, `ssh`, etc.
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/README.md:4]: done

File: services/network/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# Network
+This directory contains the network services. There is should be defined network related commands like `ping`, `traceroute`, `netstat`, etc.
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/README.md]: done

File: services/network/client/infrastructure/input/tcp-check.controller.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+			status = cliUtils.ColoredSprintf(cliUtils.RED_TEXT_COLOR, "Unexpected error - %s", result.Err.Error())
+		} else if result.Ok == true {
+			status = cliUtils.ColoredSprint(cliUtils.GREEN_TEXT_COLOT, "Succeed")
+		} else {
+			status = cliUtils.ColoredSprintf(cliUtils.RED_TEXT_COLOR, "Failed - %s", *result.FailReason)
+		}
+
+		fmt.Printf(
+			"- From target %s to %s, status: %s\n",
+			cliUtils.ColoredSprintf(cliUtils.YELLOW_TEXT_COLOR, result.Target),
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/client/infrastructure/input/tcp-check.controller.go]: Done, and check it works correctly

File: services/network/server/infrastructure/output/tcp.client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/server/infrastructure/output/tcp.client.go]: Done

File: services/network/client/application/tcp-check.usecase.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/client/application/tcp-check.usecase.go]: Dependency on any transport specific types in app/business logic force you to change app logic for it. E.g. 
- to storing `port` in go it enough uint8 and app/domain logic (unfortunately we not have it) could use this type, but protobuf not support it and I forced to use uint32. If we look into future when we would extend usecase it would force us to convert port from pb to uint8 and vise versa to use it
- This point would go even further. I'm creating GRPC's `TCPCheckRequest` in app logic and need to convert port to uint32 without any business reason. I mix real app logic and logic what appear only because of GRPC and it makes harder to understand what and why is happening. To avoid that we need to create some adapter to wrap GRPC client and introduce our own types what would be 100% suitable for business logic without unnecessary conversion

I used such approaches before and saw good results. For this particular project I'm not sure yet. I would trust you experience until I had strong enough arguments

File: services/network/client/infrastructure/input/tcp-check.controller.go (hunk 1 above)
Comment (sfc-gh-ikryvanos) [reply services/network/client/infrastructure/input/tcp-check.controller.go:70]: Let me become more familiar with a project and I'll be able to refactor it.

File: services/network/client/infrastructure/input/tcp-check.controller.go (hunk 2 above)
Comment (sfc-gh-ikryvanos) [reply services/network/client/infrastructure/input/tcp-check.controller.go:51]: Changed

File: services/network/network.proto [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+// TCP Check describes the request to make check of tcp connectivity check
+message TCPCheckRequest {
+  // The hostname to check
+  string hostname = 1;
+  // The port to check
+  uint32 port = 2;
+  // The timeout in seconds
+  uint32 timeout = 3;
```
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/network.proto]: Done

File: services/network/network.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/network.proto]: done

File: services/network/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/README.md]: done

File: services/network/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/README.md]: done

File: services/network/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ikryvanos) [reply services/network/README.md:4]: done

**Comment by sfc-gh-ikryvanos** [comment]
@sfc-gh-srhodes feel free to resolve discussions where issue being solved

File: services/network/client/application/tcp-check.usecase.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sfc-gh-srhodes) [reply services/network/client/application/tcp-check.usecase.go]: I'd rather keep the existing conventions of the codebase for now. I'm open to revisiting later on.

If we wanted to change it, we should have a solid plan for changing it throughout the codebase so that we don't end up with different styles of abstraction based on when the code was written.

(My own coding style tends to avoid abstractions and indirections until I have 2-3 concrete use cases for them. Once the use cases emerge, I refactor the code to add in the abstractions.)

**Review by sfc-gh-srhodes — COMMENTED** [review]
Could you move the license check change to a different PR? It looks logically separate from the rest.

**Comment by sfc-gh-ikryvanos** [comment]
@sfc-gh-srhodes 
>Could you move the license check change to a different PR? It looks logically separate from the rest.

Done https://github.com/Snowflake-Labs/sansshell/pull/473

**Review by sfc-gh-srhodes — APPROVED** [review]

**Review by sfc-gh-elinardi — APPROVED** [review]
